app/scripts/services/oauth/base-login-handler.ts

```diff
@@ -0,0 +1,176 @@
+import { AuthConnection } from '@metamask/seedless-onboarding-controller';
+import { OAuthErrorMessages } from '../../../../shared/modules/error';
+import { LoginHandlerOptions, AuthTokenResponse, OAuthUserInfo } from './types';
+
+export abstract class BaseLoginHandler {
+  public options: LoginHandlerOptions;
+
+  public codeVerifier: string | undefined;
+
+  // For the verification of the state (in the client side)
+  public nonce: string | undefined;
+
+  // This prompt value is used to force the user to select an account before OAuth login
+  protected readonly prompt = 'select_account';
+
+  protected readonly CODE_CHALLENGE_METHOD = 'S256';
+
+  protected readonly AUTH_SERVER_TOKEN_PATH = '/api/v1/oauth/token';
+
+  protected readonly AUTH_SERVER_REVOKE_PATH = '/api/v1/oauth/revoke';
+
+  constructor(options: LoginHandlerOptions) {
+    this.options = options;
+  }
+
+  abstract get authConnection(): AuthConnection;
+
+  abstract get scope(): string[];
+
+  /**
+   * Generate the Auth URL to initiate the OAuth login to get the Authorization Code.
+   *
+   * @returns The URL to initiate the OAuth login.
+   */
+  abstract getAuthUrl(): Promise<string>;
+
+  /**
+   * Get the JWT Token from the Web3Auth Authentication Server.
+   *
+   * @param code - The authorization code from the social login provider.
+   * @returns The JWT Token from the Web3Auth Authentication Server.
+   */
+  abstract getAuthIdToken(code?: string | null): Promise<AuthTokenResponse>;
+
+  /**
+   * Generate the request body data to get the JWT Token from the Web3Auth Authentication Server.
+   *
+   * @param code - The authorization code from the social login provider.
+   * @returns The request data for the Web3Auth Authentication Server.
+   */
+  abstract generateAuthTokenRequestData(code: string): string;
+
+  /**
+   * Get the user's information from the JWT Token.
+   *
+   * @param idToken - The JWT Token from the Web3Auth Authentication Server.
+   * @returns The user's information from the JWT Token.
+   */
+  abstract getUserInfo(idToken: string): Promise<OAuthUserInfo>;
+
+  /**
+   * Validate the state value from the OAuth login redirect URL.
+   *
+   * @param state - The state value from the OAuth login redirect URL.
+   */
+  validateState(state: unknown): void {
+    if (typeof state !== 'string') {
+      throw new Error(OAuthErrorMessages.INVALID_OAUTH_STATE_ERROR);
+    }
+
+    const parsedState = JSON.parse(state);
+
+    if (parsedState.nonce !== this.nonce) {
+      throw new Error(OAuthErrorMessages.INVALID_OAUTH_STATE_ERROR);
+    }
+  }
+
+  /**
+   * Refresh the JWT Token using the refresh token.
+   *
+   * @param refreshToken - The refresh token from the Web3Auth Authentication Server.
+   * @returns The JWT Token from the Web3Auth Authentication Server and new refresh token.
+   */
+  async refreshAuthToken(refreshToken: string): Promise<AuthTokenResponse> {
```

Comment:
I prefer to have refreshAuthToken , revokeRefreshToken and requestAuthToken function to be a different class or object